File: Danil_Safronov/FIB.scala
import scala.annotation.tailrec
import scala.io.StdIn.readInt

object FIB {
  def main(args: Array[String]) {
    print(countRabbits(readInt, readInt))
  }

  def countRabbits(n: Int, k: Int): Long = {
    @tailrec def updRabbits(n: Int, k: Int, adult: Long, small: Long): Long = n match {
      case 0 => 1
      case 1 => adult + small
      case _ => updRabbits(n - 1, k, adult + small, adult * k)
    }

    updRabbits(n, k, 0, 1)
  }
}

File: lev_karavashkin/src/main/java/FIB.java
public class FIB {
    public static int getTotalRabbitPairs(int n, int k) {

        if(n >= 40 || k > 5){
            return -1;
        }

        int[] F =  new int[n];
        F[0] = 1;
        F[1] = 1;


        for (int i = 2; i < n; i++) {
            F[i] = F[i-1] + F[i-2] * k;
        }

        return F[n-1];
    }
}
